Mathlib/RepresentationTheory/Coinvariants.lean

```diff
@@ -0,0 +1,405 @@
+noncomputable def finsuppToCoinvariants :
+    (α →₀ Coinvariants ρ) →ₗ[k] Coinvariants (ρ.finsupp α) :=
+  lsum (R := k) k fun a => Coinvariants.lift _ (Coinvariants.mk _ ∘ₗ lsingle a) fun g =>
+    LinearMap.ext fun x => (mk_eq_iff _).2 <|
+    mem_augmentationSubmodule_of_eq g (single a x) _ <| by simp
+
+@[simp]
+lemma finsuppToCoinvariants_single_mk (a : α) (x : V) :
+    finsuppToCoinvariants ρ α (single a <| Coinvariants.mk _ x) =
+      Coinvariants.mk _ (single a x) := by
+  simp [finsuppToCoinvariants]
+
+variable (ρ α) in
+/-- Given a `G`-representation `(V, ρ)` and a type `α`, this is the linear equivalence
+`(α →₀ V)_G ≃ₗ (α →₀ V_G)` sending `⟦single a v⟧ ↦ single a ⟦v⟧`. -/
+@[simps! symm_apply]
+noncomputable def coinvariantsFinsuppLEquiv :
+    Coinvariants (ρ.finsupp α) ≃ₗ[k] α →₀ Coinvariants ρ :=
+  LinearEquiv.ofLinear (coinvariantsToFinsupp ρ α) (finsuppToCoinvariants ρ α)
+    (by ext; simp) (by ext; simp)
+
+@[simp]
+lemma coinvariantsFinsuppLEquiv_apply (x) :
+    coinvariantsFinsuppLEquiv ρ α x = coinvariantsToFinsupp ρ α x := by rfl
+
+end Finsupp
+
+section TensorProduct
+
+open TensorProduct Coinvariants Finsupp
+
+variable {k G V W : Type*} [CommRing k] [Group G] [AddCommGroup V] [Module k V]
+  [AddCommGroup W] [Module k W] (ρ : Representation k G V) (τ : Representation k G W)
+
+@[simp]
+lemma Coinvariants.mk_inv_tmul (x : V) (y : W) (g : G) :
+    Coinvariants.mk (ρ.tprod τ) (ρ g⁻¹ x ⊗ₜ[k] y) = Coinvariants.mk (ρ.tprod τ) (x ⊗ₜ[k] τ g y) :=
+  (mk_eq_iff _).2 <| mem_augmentationSubmodule_of_eq g⁻¹ (x ⊗ₜ[k] τ g y) _ <| by simp
+
+@[simp]
+lemma Coinvariants.mk_tmul_inv (x : V) (y : W) (g : G) :
+    Coinvariants.mk (ρ.tprod τ) (x ⊗ₜ[k] τ g⁻¹ y) = Coinvariants.mk (ρ.tprod τ) (ρ g x ⊗ₜ[k] y) :=
+  (mk_eq_iff _).2 <| mem_augmentationSubmodule_of_eq g⁻¹ (ρ g x ⊗ₜ[k] y) _ <| by simp
+
+/-- Given a `k`-linear `G`-representation `V, ρ`, this is the map `(V ⊗ k[G])_G →ₗ[k] V` sending
+`⟦v ⊗ single g r⟧ ↦ r • ρ(g⁻¹)(v)`. -/
+noncomputable def ofCoinvariantsTprodLeftRegular :
+    Coinvariants (ρ.tprod (leftRegular k G)) →ₗ[k] V :=
+  Coinvariants.lift _ (TensorProduct.lift (Finsupp.linearCombination _ fun g => ρ g⁻¹) ∘ₗ
+    (TensorProduct.comm _ _ _).toLinearMap) fun _ => by ext; simp
+
+@[simp]
+lemma ofCoinvariantsTprodLeftRegular_mk_tmul_single (x : V) (g : G) (r : k) :
+    ofCoinvariantsTprodLeftRegular ρ (Coinvariants.mk _ (x ⊗ₜ Finsupp.single g r)) = r • ρ g⁻¹ x :=
+  congr($(Finsupp.linearCombination_single k (v := fun g => ρ g⁻¹) r g) x)
+
+/-- Given a `k`-linear `G`-representation `(V, ρ)`, this is the linear equivalence
+`(V ⊗ k[G])_G ≃ₗ[k] V` sending `⟦v ⊗ single g r⟧ ↦ r • ρ(g⁻¹)(v)`. -/
+@[simps! symm_apply]
+noncomputable def coinvariantsTprodLeftRegularLEquiv :
+    Coinvariants (ρ.tprod (leftRegular k G)) ≃ₗ[k] V :=
+  LinearEquiv.ofLinear (ofCoinvariantsTprodLeftRegular ρ)
+    (Coinvariants.mk _ ∘ₗ (TensorProduct.mk k V (G →₀ k)).flip (single 1 1)) (by ext; simp) <| by
+      ext x g
```

Comment:
Why doesn't `ext; simp` work? Is there a missing `ext` lemma?

Reply:
ooh, `ext; simp` does work